### PR TITLE
Add fusion enabled apis

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#64429 Add fusion enabled apis**
* #64428 update optimize_for_inference docs

Differential Revision: [D30740897](https://our.internmc.facebook.com/intern/diff/D30740897)

### CODE REVIEW

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -197,6 +198,34 @@ def _hide_source_ranges() -> Iterator[None]:
...
+
+    if device is None:
+        torch._C._jit_override_can_fuse_on_cpu(enabled)
+        torch._C._jit_override_can_fuse_on_gpu(enabled)
+    else:
+        assert device in ["cpu", "cuda"], "Device-specific fusion must be equal to 'cpu' or 'cuda' if not None"
+        if device == "cpu":
+            torch._C._jit_override_can_fuse_on_gpu(enabled)
+        else:
+            torch._C._jit_override_can_fuse_on_cpu(enabled)
```
Comment (ZolotukhinM) [inline torch/jit/__init__.py:218]: If-else statements seem swapped.

**Review by ZolotukhinM — APPROVED** [review]
Nice!

**Review by bertmaher — APPROVED** [review]
Awesome, thank you!

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30740897).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30740897).

**Pull request closed without merging**